File: python/coinbase-agentkit/coinbase_agentkit/action_providers/action_policy.py
"""Optional policy hooks for AgentKit action execution."""
from dataclasses import dataclass
from typing import Any, Protocol

from ..wallet_providers import WalletProvider

@dataclass(frozen=True)
class ActionPolicyContext:
    """Context passed to policy hooks for one action invocation."""
    action_name: str
    provider_name: str
    wallet_provider: WalletProvider
    args: Any

class ActionPolicy(Protocol):
    """Provider-agnostic hooks for approval, budgets, auditing, or release."""
    def before_action(self, context: ActionPolicyContext) -> bool | str: ...
    def after_action(self, context: ActionPolicyContext, result: str) -> None: ...
    def on_action_error(self, context: ActionPolicyContext, error: BaseException) -> None: ...

class ActionPolicyDeniedError(RuntimeError):
    """Raised when a policy prevents an action from executing."""
    def __init__(self, action_name: str, reason: str | None = None) -> None:
        message = f"Action {action_name!r} denied"
        if reason:
            message += f": {reason}"
        super().__init__(message)
        self.action_name = action_name
        self.reason = reason

File: typescript/agentkit/src/action-providers/actionPolicy.ts
import { WalletProvider } from "../wallet-providers";

/** Context supplied to a policy before and after an action runs. */
export type ActionPolicyContext = {
  actionName: string;
  providerName: string;
  walletProvider: WalletProvider;
  args: unknown;
};

/** Result returned by a policy pre-flight check. */
export type ActionPolicyDecision = {
  allowed: boolean;
  reason?: string;
};

/** Optional, provider-agnostic guard around action execution. */
export interface ActionPolicy {
  beforeAction?: (
    context: ActionPolicyContext,
  ) => ActionPolicyDecision | Promise<ActionPolicyDecision>;
  afterAction?: (
    context: ActionPolicyContext,
    result: string,
  ) => void | Promise<void>;
  onActionError?: (
    context: ActionPolicyContext,
    error: unknown,
  ) => void | Promise<void>;
}

/** Error raised when an installed policy denies an action. */
export class ActionPolicyDeniedError extends Error {
  public readonly actionName: string;
  public readonly reason?: string;

  constructor(actionName: string, reason?: string) {
    super(reason ? `Action "${actionName}" denied: ${reason}` : `Action "${actionName}" denied by policy`);
    this.name = "ActionPolicyDeniedError";
    this.actionName = actionName;
    this.reason = reason;
  }
}
